### PR TITLE
New artifacts category for corefxlab

### PR DESCRIPTION
**Pull request by pgovind**
Part of moving [corefxlab](https://github.com/dotnet/corefxlab) to arcade.

### CODE REVIEW

**Comment by mmitche** [comment]
I don't think we should do this this way. We intend to remove this code eventually.  Why not just associate corefxlab with the .NET 5 Dev channel?

**Comment by pgovind** [comment]
@safern can answer that question much better than me.

**Comment by safern** [comment]
> Why not just associate corefxlab with the .NET 5 Dev channel?

The intention of corefxlab is to have experimental packages that potentially can be moved to the product itself, so I wouldn't publish to the same feed as .NETCORE to avoid potential package clashes in the future.

**Comment by riarenas** [comment]
I agree with @mmitche that this smells fishy. 

To which channel are corefx builds going to publish? depending on the channel the builds will end up in azure devops feeds regardless of the artifact category, which we are trying to move away from.

The corefxlab build could publish to the `general testing` channel or if for some reason that's not enough, we could create a channel specifically for this.

Also, the right path may vary depending on how the packages will become available to the product.
* Is the model here that once a package is ready, the code will be moved to another repo, and thus the package will be produced by its new home?
* Is the model instead that corefxlab will always produce the packages, but only certain versions need to be published to the feeds that other parts of the product consume?

I *think* what @safern means is option 1, but would like some clarification.

**Comment by mmitche** [comment]
> > Why not just associate corefxlab with the .NET 5 Dev channel?
> 
> The intention of corefxlab is to have experimental packages that potentially can be moved to the product itself, so I wouldn't publish to the same feed as .NETCORE to avoid potential package clashes in the future.

What kind of package clashes are you worried about? Btw, since all of your packages will initially be non-shipping,. they're actually go to dotnet5-transport and will have a full non-stable suffix.

I do not want to introduce extra feeds. The systems we have in place are already complex and difficult to reason about. I would just publish to the regular .net 5 channel. When packages move to corefx, chances of exact version collisions are extremely small (only possible on the same day as the move).

**Comment by riarenas** [comment]
Maybe that is even overkill. @safern is the concern only in package version clashes? that seems like an unlikely scenario unless you plan to produce stable versions of the packages in corefxlab, which would seem to contradict the experimental nature of corefxlab in the first place.

**Comment by safern** [comment]
> What kind of package clashes are you worried about? Btw, since all of your packages will initially be non-shipping,. they're actually go to dotnet5-transport and will have a full non-stable suffix.

> I do not want to introduce extra feeds. The systems we have in place are already complex and difficult to reason about. I would just publish to the regular .net 5 channel. When packages move to corefx, chances of exact version collisions are extremely small (only possible on the same day as the move).

I spoke with @riarenas offline about this. So corefxlab has always been weird in versioning were some packages are 2.0.0, some 1.0.0, some 0.1.0 and so on. This is because we shipped this packages previously and for some packages we needed to bump the major or minor version for sorting and to follow nuget guidelines where the nuget version should be higher. 

Also, the suffix we where using at some time was not sorting correctly so we fixed it to something else. 

Corefxlab is a place where we experiment, or bring back old packages that might be interesting and if they are successful then we could move them to corefx and ship them as a product package. That's why we've always published to a corefxlab feed in myget. This to avoid having a package been 1.0.0 or 2.0.0 or whatever version in corefxlab feed and then when we move it to corefx, they could downgrade and show as older packages to external customers. Why downgrade? Well it can still be experimental in corefx so we want to probably version it 0.1.0 then 0.2.0 and so on until we're ready to ship stable. So I would really want to have a customer experience were there is a public feed for this packages, as experimental packages and so that they use them specifically from a different feed knowing they're experimental. 

> that seems like an unlikely scenario unless you plan to produce stable versions of the packages in corefxlab, which would seem to contradict the experimental nature of corefxlab in the first place.

We're actually producing a 0.1.0 stable version of one package now and published to NuGet a few weeks ago as we want official products to use it to experiment with it and see perf gain and other measurments to decide if we should move it to dotnet/runtime. 

All this comes because we want to start following arcade versioning and using arcade toolsets, so we decided to kill myget. I would suggest the best way to go (maybe not from engineering, but customer point of view) is to add something like a .NET Experimental channel where probably in the future other products could publish to it.

**Comment by riarenas** [comment]
I can see a use case for an "experimental" category (so that packages don't go to dotnet-core by default while we remove the categories functionality) and a channel + azdo feed that goes with it, just not sure if there are actually more uses for this, or if it would end up becoming just a way to deal with the versioning quirks of the corefxlab packages.

@mmitche Do you think it would be worth introducing such a channel/feeds for scenarios similar to this?

**Comment by mmitche** [comment]
So that makes it sounds like it should be another channel.  Updating the artifact category is not the correct way to do this, anyway.  I'll file an issue to make an experimental channel.

The one thing that gives me pause is that there does seem to be complexity for complexity's sake. If the packages are experimental, why did they ever make it past 0.x? A lot of these decisions asking for additional feeds, channels, etc. look fine on an individual basis, but the total sum cost has been a harder and harder to understand repository ecosystem.

**Comment by mmitche** [comment]
> I can see a use case for an "experimental" category (so that packages don't go to dotnet-core by default while we remove the categories functionality) and a channel + azdo feed that goes with it, just not sure if there are actually more uses for this, or if it would end up becoming just a way to deal with the versioning quirks of the corefxlab packages.
> 
> @mmitche Do you think it would be worth introducing such a channel/feeds for scenarios similar to this?

Yeah, can you file an issue to do this? (or just do it)

**Comment by riarenas** [comment]
@mmitche I think a version of this change might still be required so that the packages don't end up in the dotnet-core blob feed and instead get their own isolated feed? 

Maybe just replace the use of "corefxlab" with "experimental"?

**Comment by mmitche** [comment]
> @mmitche I think a version of this change might still be required so that the packages don't end up in the dotnet-core blob feed and instead get their own isolated feed?
> 
> Maybe just replace the use of "corefxlab" with "experimental"?

I suppose that's okay until we get rid of blob feeds.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SetupTargetFeeds.proj [hunk 1]
```diff
@@ -122,6 +122,7 @@
       <TargetStaticFeed Condition="'$(ArtifactsCategory.ToUpper())' == 'ASPNETENTITYFRAMEWORK6'">https://dotnetfeed.blob.core.windows.net/aspnet-entityframework6/index.json</TargetStaticFeed>
       <TargetStaticFeed Condition="'$(ArtifactsCategory.ToUpper())' == 'ASPNETBLAZOR'">https://dotnetfeed.blob.core.windows.net/aspnet-blazor/index.json</TargetStaticFeed>
       <TargetStaticFeed Condition="'$(ArtifactsCategory.ToUpper())' == 'IOT'">https://dotnetfeed.blob.core.windows.net/dotnet-iot/index.json</TargetStaticFeed>
+      <TargetStaticFeed Condition="'$(ArtifactsCategory.ToUpper())' == 'COREFXLAB'">https://dotnetfeed.blob.core.windows.net/dotnet-corefxlab/index.json</TargetStaticFeed>
```
Comment (riarenas) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SetupTargetFeeds.proj]:
```suggestion
      <TargetStaticFeed Condition="'$(ArtifactsCategory.ToUpper())' == 'EXPERIMENTAL'">https://dotnetfeed.blob.core.windows.net/dotnet-experimental/index.json</TargetStaticFeed>
```

**Pull request closed without merging**